### PR TITLE
MongoCollection Unit Tests

### PR DESCRIPTION
**Pull request by tigerstrikemedia**
Added setup requirements to docker-compose and an initial set of unit tests for the MongoCollection.

### CODE REVIEW

**Comment by tigerstrikemedia** [comment]
What did I mess up so that the @expectedException in an unrelated test isn't catching the exception that is correctly thrown and killing the test run?

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 1]
```diff
@@ -0,0 +1,334 @@
+<?php
+
+use Phalcon\Mvc\Model\Message as ModelMessage;
+use Phalcon\Mvc\MongoCollection;
+use Phalcon\Db\Adapter\MongoDB\Client as MongoClient;
+
+class CollectionsTest extends PHPUnit_Framework_TestCase
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: First of all  use please Codeception tests:
```php
<?php

namespace Phalcon\Test\Mvc;

use Codeception\TestCase\Test;

class CollectionsTest extends extends Test
{
    // ...
}
```

See for example: `tests/unit/Validation/Validator/MongoIdTest.php`

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    public function collectionsAutoloader($className)
+    {
+        if (file_exists('tests/unit/collections/' . $className . '.php')) {
+            require 'tests/unit/collections/' . $className . '.php';
+        }
+    }
+
+    public function setUp()
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]:
```php
    protected function _before()
    {
        if (!extension_loaded('mongo')) {
            $this->markTestSkipped(
                'The Mongo extension is not available.'
            );
        }
    }
```

File: tests/build.sh [hunk 3]
```diff
@@ -73,6 +73,5 @@ ${docker_bin} run -it --rm \
   -v ${TRAVIS_BUILD_DIR}/codeception.yml:/app/codeception.yml \
   -v ${TRAVIS_BUILD_DIR}/tests:/app/tests \
   -v ${TRAVIS_BUILD_DIR}/Library:/app/Library \
-  -v ${TRAVIS_BUILD_DIR}/vendor:/app/vendor \
```
Comment (sergeyklay) [inline tests/build.sh]: Why? Here is Composer autoloader. Codeception use it

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+class CollectionsTest extends PHPUnit_Framework_TestCase
+{
+    public function __construct()
+    {
+        spl_autoload_register(array($this, 'collectionsAutoloader'));
+    }
+
+    public function __destruct()
+    {
+        spl_autoload_unregister(array($this, 'collectionsAutoloader'));
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: Not needed. Just use Composer autoloader from vendor and namespaces. See: `composer.json` `"autoload-dev"`

File: tests/unit/collections/Cars.php [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+
```
Comment (sergeyklay) [inline tests/unit/collections/Cars.php]: Jut use here namespace

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+        if( ! extension_loaded( 'MongoDB' ) ){
+            $this->markTestSkipped("MongoDB extension not loaded, test skipped");
+            return;
+        }
+
+        Phalcon\DI::reset();
+        $di = new Phalcon\DI();
+        $di->set('mongo', function(){
+            $mongo = new MongoClient( 'mongodb://' . TEST_MONGODB_HOST . ':27017');
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: Add please port to the constant plase

File: tests/build.sh (hunk 3 above)
Comment (tigerstrikemedia) [inline tests/build.sh]: It is a duplicate of a line before it. 

line 72:
`-v ${TRAVIS_BUILD_DIR}/vendor:/app/vendor \`

Happy to add it back in if it is needed.

Comment (sergeyklay) [inline tests/build.sh]: Hm.. Okay

**Comment by tigerstrikemedia** [comment]
@sergeyklay its 1:37am here, let me know if there is anything else i need to cover in the morning ;)

Cheers
Ben

File: tests/_bootstrap.php [hunk 7]
```diff
@@ -57,3 +57,6 @@
 define('TEST_DB_PASSWD', getenv('TEST_DB_PASSWD') ?: '');
 define('TEST_DB_NAME', getenv('TEST_DB_NAME') ?: 'incubator_tests');
 define('TEST_DB_CHARSET', getenv('TEST_DB_CHARSET') ?: 'utf8');
+
+// MongoDB
+define('TEST_MONGODB_HOST', getenv('TEST_MONGODB_HOST') ?: 'mongodb' );
```
Comment (sergeyklay) [inline tests/_bootstrap.php]:
```php
define('TEST_MONGODB_PORT', getenv('TEST_MONGODB_PORT') ?: '27017' );
```

Comment (sergeyklay) [inline tests/_bootstrap.php]: Also add new constants to the `docs/TESTING.md`

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+
+        if( ! extension_loaded( 'MongoDB' ) ){
+            $this->markTestSkipped("MongoDB extension not loaded, test skipped");
+            return;
+        }
+
+        Di::reset();
+        $di = new DI();
+        $di->set('mongo', function(){
+            $mongo = new MongoClient( 'mongodb://' . TEST_MONGODB_HOST . ':27017');
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: TEST_MONGODB_PORT

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        $car->model = '650S';
+        $car->rank = 1;
+        $car->value = 500000;
+        $car->save();
+
+        $car = new Cars();
+        $car->manufacturer = 'Porsche';
+        $car->model = '911 GT3';
+        $car->rank = 2;
+        $car->value=450000;
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: Could you please use common code style? Before `$car->value = 500000;`, but here `$car->value=450000;`

File: tests/unit/Mvc/MongoCollectionTest.php [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+use Codeception\TestCase\Test;
+use Phalcon\Test\Collections\Cars;
+
+class CollectionsTest extends Test
+{
+
+    protected function _before()
+    {
+
+        if( ! extension_loaded( 'MongoDB' ) ){
```
Comment (sergeyklay) [inline tests/unit/Mvc/MongoCollectionTest.php]: Could you please follow PSR code style?
```php
if (!extension_loaded('MongoDB')) {
```

**Comment by sergeyklay** [comment]
@tigerstrikemedia Good work!

**Comment by sergeyklay** [comment]
Also could you please rebase. See https://github.com/phalcon/incubator/pull/639

File: tests/unit/collections/Cars.php [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<?php
+namespace Phalcon\Test\Collections;
+
+class Cars extends \Phalcon\Mvc\MongoCollection{
+}
```
Comment (sergeyklay) [inline tests/unit/collections/Cars.php]: Fine! I suggest to move this file for example to `tests/_data/collections` with `Phalcon\Test\Collections` and amend composer.json:

```js
    "autoload": {
        "psr-4": {
            "Phalcon\\Test\\": "tests/unit/",
            "Phalcon\\Test\\Collections\\": "tests/_data/collections/"
        }
    }
```

Also cpuld you please follow PSR:

```php
class Cars extends \Phalcon\Mvc\MongoCollection
{
```

File: Library/Phalcon/Mvc/MongoCollection.php [hunk 12]
```diff
@@ -2,35 +2,38 @@
...
+ * Tested Using Unit Tests:
+ * save()
  * delete()
+ * find()
+ * count()
+ * findById()
+ * findFirst()
+ * aggregate()
+ *
+ * Tests To Write
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/MongoCollection.php]: I suggest to move all test-related info and todo to the `docs/TESTING.md`

**Comment by sergeyklay** [comment]
@tigerstrikemedia It seems to me all ok. Remains `tests/unit/Validation/Validator/Db/UniquenessTest.php:110`  - it should be updated in accordance with Phalcon's new validation logic. Could you please fix it?

**Comment by sergeyklay** [comment]
@tigerstrikemedia See `Library/Phalcon/Validation/Validator/Db/Uniqueness.php:78` at 2.1.x branch

**Comment by tigerstrikemedia** [comment]
@sergeyklay 

`Fatal error:  __debuginfo() must return an array in /app/Library/Phalcon/Validation/Validator/Db/Uniqueness.php on line 79`

I have no idea where this little gem comes from, from what I can see it is killing the test run.

**Comment by sergeyklay** [comment]
Ya, it seems it requires DB connection

**Comment by tigerstrikemedia** [comment]
@sergeyklay 

```
/**
  * @expectedException        \Phalcon\Validation\Exception
  * @expectedExceptionMessage Validator Uniquness require connection to database
  */
```

Why is that triggering a cryptic PHP fatal error?

**Comment by sergeyklay** [comment]
@tigerstrikemedia No. The `testAvailableUniquenessWithDefaultDI` throws exception.

**Comment by sergeyklay** [comment]
@tigerstrikemedia Try in test `Di::setDefault($di);` before `new Uniqueness($uniquenessOptions);` or something like that

**Comment by tigerstrikemedia** [comment]
Where are you seeing that it crashes at `testAvailableUniquenessWithDefaultDI ` ?

https://travis-ci.org/phalcon/incubator/jobs/141949576

Starts with a PHP warning here: 
`- UniquenessTest: Should catch exception when validate uniquness without db and default di`

and crashes with a fatal error: 
`__debuginfo() must return an array in /app/Library/Phalcon/Validation/Validator/Db/Uniqueness.php on line 79`

Which is exactly what is expected to happen. 
`throw new ValidationException('Validator Uniquness require connection to database');`

**Comment by sergeyklay** [comment]
screenshot at jul 04 01-34-42

**Comment by sergeyklay** [comment]
Thanks

**Comment by tigerstrikemedia** [comment]
np @sergeyklay thanks for pulling me up on sloppy work ;)

**Pull request merged**